### PR TITLE
Apple Universal Link & Android App Link Support via Interapp

### PR DESCRIPTION
**Pull request by AndrewEQ**
https://developer.apple.com/ios/universal-links

https://developer.android.com/training/app-links

Todo: Add usage to Interapp docs

### CODE REVIEW

File: lib/UnoCore/Targets/Android/com/fuse/App.java [hunk 1]
```diff
@@ -134,7 +134,6 @@ public void onResume() {
         ActivityNativeEntryPoints.cppOnResume();
         com.fuse.Activity.onResume();
         focusOnResume();
-
```
Comment (kusma) [inline lib/UnoCore/Targets/Android/com/fuse/App.java]: This looks like an unneeded change. Please drop it.

File: lib/UnoCore/Targets/Android/com/fuse/App.java [hunk 2]
```diff
@@ -217,6 +216,11 @@ public void onNewIntent (Intent intent)
             pendingURI = intent.getDataString();
         }
 #endif
+
+#if @(Project.Android.AssociatedDomains:IsSet)
+
+        @(Project.Android.AssociatedDomains:SplitAndJoin('\n\t\t', 'if (intent!=null && intent.getScheme()!=null && intent.getScheme().contains("http") && intent.getDataString().toLowerCase().contains(new String("','").toLowerCase())) { pendingURI = intent.getDataString(); }'))
```
Comment (kusma) [inline lib/UnoCore/Targets/Android/com/fuse/App.java]: Is there a reason for the SplitAndJoin-stuff here? Can't you just put the code in there verbatim instead?

File: lib/UnoCore/Targets/Android/com/fuse/App.java (hunk 1 above)
Comment (AndrewEQ) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: Haha, I knew you would say that

File: lib/UnoCore/Targets/Android/com/fuse/App.java (hunk 2 above)
Comment (AndrewEQ) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: Needed to cater for multiple domains in the .unoproj:
```
  "Android": {
    "AssociatedDomains": ["example.com","sub.example.com"],
   ...
```

File: lib/UnoCore/Targets/Android/com/fuse/App.java (hunk 1 above)
Comment (AndrewEQ) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: Sorted.

File: lib/UnoCore/Targets/Android/com/fuse/App.java (hunk 2 above)
Comment (AndrewEQ) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: Other syntax ideas for handling multiple?

Comment (kusma) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: How about generating an array of strings, and avoid repeating the code for each domain?

Comment (AndrewEQ) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: Oh yeah, I guess my mind got stuck in markup land... sorted.

File: lib/UnoCore/Targets/Android/com/fuse/App.java [hunk 3]
```diff
@@ -208,6 +208,7 @@ public void onActivityResult (int arg0, int arg1, android.content.Intent arg2)
     //------------------------------------------------------------
 
     private String pendingURI = "";
+    private String[] associatedDomains;
```
Comment (kusma) [inline lib/UnoCore/Targets/Android/com/fuse/App.java]: I don't think you need a member-variable for this, a local variable should do just fine, no?

Comment (AndrewEQ) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: I guess so, just thought if someone else needed it in the future they could use it but sure, lemme localise it

Comment (AndrewEQ) [reply lib/UnoCore/Targets/Android/com/fuse/App.java]: Ok, sorted ;)

**Comment by AndrewEQ** [comment]
Ha, thanks! 

In a nutshell, when someone taps on a web link in WhatsApp for example, it will open your app up and pass the url in, same behaviour for android.

**Comment by kusma** [comment]
Ah, so more than just custom Uri schemes, like we already support? That's pretty cool :)

**Comment by AndrewEQ** [comment]
Yeah, its more than the custom Uri scheme support; this is essentially binding the app to a domain that you have to verify, I routed the event through the existing InterApp modules receiving uri event to make things easier and well, code reuse.

So custom uri's will be like so: myapp://dosomething whereas, this is http://mydomain.com/dosomething

People don't exactly share urls like "myapp://dosomething" cos if the app is not installed, there's no fallback user experience, whereas with a url, it will just open up a browser. I think for android, if you used custom scheme to make a url, the difference is that android would then bring up a menu option asking you which app you would like to open the url with, whereas with "app links", they automatically just open the app 😄

**Pull request merged**